### DISCUSSION TITLE
wrong result and convergence issues for single-phase 1D flow

### DISCUSSION THREAD

**Discussion by Raven-pro** [idaholab/moose#22682] [Q&A General] [answered]
Hi, I am opening a new topic from the previous thread [#22648]. The previous problem is about implementing AD to the fp coefficients so that the jacobian shall be formed properly, and I tried, but it turned out not so many differences were made on my issues.
I am trying to build a single-phase 1D 3Eqns flow application based on MOOSE, the governing equations are as follows.<img width="330" alt="截屏2022-11-13 13 51 53" src="https://user-images.githubusercontent.com/68334741/201507976-2ae0cd0c-7320-4de6-8a07-e2a3189aa0cc.png">
And transforming the equations to the weak forms, with the divergence theorem, I have
<img width="658" alt="截屏2022-11-14 20 19 54" src="https://user-images.githubusercontent.com/68334741/201658511-ff3e55a8-8ebc-4bad-b8b6-357e4d212d2e.png">
Then I wrote kernels and BCS based on the above weak-formed equations, with ICs and BCs, set as in the input file below.
```

    p_init = 15.5e6       # Pa
    h_init = 1133820    # J/kg 
    v_init = 1                 # m/s
    De = 0.025             # m
    U = 0.0785398      # m
    A = 0.00049087    # m2
    q = 1273239.545   # W/m2


[Mesh]
    type = GeneratedMesh
    dim = 1
    nx = 50
    # ny = 12
    # ymax = 0.012
    # ymin = 0
    xmax = 1.0
    xmin = 0
    elem_type = EDGE
[]

[Modules]
    [FluidProperties]
        [water]
            type = Water97FluidProperties
        []
    []
[]

[Materials]
    [fp_mat]
        type = Water97
        p = P
        h = H
        fluid = water
    []
[]

[Variables]
    [P]
        order = FIRST
        family = LAGRANGE
        scaling = 1
        initial_condition = ${p_init}
    []
    [V]
        order = FIRST
        family = LAGRANGE
        scaling = 1e-3
        initial_condition = ${v_init}
    []
    [H]
        order = FIRST
        family = LAGRANGE
        scaling = 1e-6
        initial_condition = ${h_init}
    []
[]


[Kernels]
    [MassTime]
        type = MassTime
        variable = P
        dot_h = H
    []
    [MassAdvection]
        type = MassAdvection
        variable = P
        v = V
    []
    [MassPSPG]
        type = MassPSPG
        variable = P
        dot_h = H
        v = V
        grad_v = V
        De = ${De}
    []
    [MomentumTime]
        type = MomentumTime
        variable = V
        dot_p = P
        dot_h = H
    []
    [MomentumAdvection]
        type = MomentumAdvection
        variable = V
    []
    [MomentumSource]
        type = MomentumSource
        variable = V
        grad_p = P
        De = ${De}
    []
    [MomentumSUPG]
        type = MomentumSUPG
        variable = V
        grad_p = P
        De = ${De}
    []
    [EnergyTime]
        type = EnergyTime
        variable = H
        dot_p = P
    []
    [EnergyAdvection]
        type = EnergyAdvection
        variable = H
        v = V
    []
    [EnergySource]
        type = EnergySource
        variable = H
        U = ${U}
        A = ${A}
        q = ${q}
    []
    [EnergySUPG]
        type = EnergySUPG
        variable = H
        v = V
        U = ${U}
        A = ${A}
        q = ${q}
    []
[]

[BCs]
    [bc1]
        type = DirichletBC
        variable = H
        value = ${h_init}
        boundary = "left"
    []
    [bc2]
        type = DirichletBC
        variable = P
        value = ${p_init}
        boundary = "right"
    []
    [bc3]
        type = DirichletBC
        variable = V
        value = ${v_init}
        boundary = "left"
    []
    [bc4]
        type = MassBC
        variable = P
        v = V
        boundary = "right"
    []
    [bc5]
        type = MomentumBC
        variable = V
        boundary = "right"
    []
    [bc6]
        type = EnergyBC
        variable = H
        v = V
        boundary = "right"
    []
[]

#[VectorPostprocessors]
  #[H]
    #type = NodalValueSampler
    #variable = H
    #boundary = 'bottom'
    #block = 0
    #execute_on = 'initial timestep_end' 
    #use_displaced_mesh = true
    #sort_by = x
  #[]
  #[P]
    #type = NodalValueSampler
    #variable = P
    #boundary = 'bottom'
    #block = 0
    #execute_on = 'initial timestep_end' 
    #use_displaced_mesh = true
    #sort_by = x
  #[]
#[]

[Problem]
    type = FEProblem
    coord_type = XYZ
[]

[Preconditioning]
    [./SMP]
        type = SMP
        full = true
        solve_type = 'PJFNK'
        petsc_options_iname = '-pc_type'
        petsc_options_value = 'lu'
    [../]
[]

[Executioner]
    type = Transient
    solve_type =  'PJFNK'
    # compute_scaling_once = false
    automatic_scaling = true
    scheme = bdf2
    dt = 1e-5
    dtmin = 1e-5
    petsc_options_iname = '-ksp_gmres_restart'
    petsc_options_value = '100'
    nl_rel_tol = 1e-7
    nl_abs_tol = 1e-6
    nl_max_its = 20

    l_tol = 1e-4
    l_max_its = 100

    start_time = 0.0
    num_steps = 200
    end_time = 1e-3

    [./Quadrature]
        type = TRAP
        order = FIRST
    [../]
[]

[Outputs]
    exodus = true
    csv = true
[]
```
But the results are wrong in the ParaView. If the code runs longer, it will not converge, the results are like that, we have negative velocity values intermittently, and the enthalpy oscillated, which is obviously wrong.
<img width="996" alt="截屏2022-11-13 14 00 59" src="https://user-images.githubusercontent.com/68334741/201508195-d07ccc49-2b41-4f95-81f4-10bfb8633c08.png">
I tried adding the PSPG term and SUPG term, but it didn't work. Can anybody knows how to make it converge? I will really appreciate it.
The code of my application is uploaded below (Mostly several kernels and three IntegratedBCs), can anybody check it a little? THX!
[SF](https://github.com/89018580/SF)

**Comment by Raven-pro** [comment, marked as answer]
I have just tested the use of stabilization and it turned out useful in stabilizing the oscillation of pressure. But oscillations of enthalpy and velocity are still unsolved.

What's more, if I set the inlet pressure boundary condition instead of the outlet, the pressure will oscillate as well.

So how to get rid of this kind of abnormal oscillation?

**Reply by GiudGiud** [reply]
Hello

so the nonlinear solves are converging but the result is oscillating unphysically?

if you are hitting convergence issues, you will want to check the Jacobian for accuracy
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

Guillaume

**Reply by Raven-pro** [reply]
Yes, when I set d_t to be 1e-5 or lower. running for like 200 steps, the nonlinear solves converge with oscillating results.

If the code runs longer for like 300 steps, the solves will not converge.  I don't know if it truly is a converging issues since the code can converge at first several hundreds of steps.

**Comment by Raven-pro** [comment]
I don't know if it's because I have missed something about MOOSE.

I have read papers about SAM, which is also based on MOOSE and can solve single-phase compressible problems, and it implements the same three governing equations in the applications, which runs well.

So I think this kind of problem maybe is because of my limited knowledge of MOOSE. 

I am just thinking, If it's really about jacobian, is it better to hand-code jaboian than AD?

**Reply by GiudGiud** [reply]
It's not better to hand code the Jacobian. The AD Jacobian is numerically perfect. If you are using AD the script I linked to should have reported a perfect Jacobian

However, you are using PJFNK which does not fully leverage the potential of AD. I recommend you try Newton's method instead, which will benefit from these accurate Jacobians from AD. 

You may want to tighten the linear tolerances and increase the order of the quadrature as well, I m not sure

**Reply by Raven-pro** [reply]
Thank you, I will try your suggestions!

**Reply by GiudGiud** [reply]
hello

where do we stand on this?
where you able to get a non-oscillating profile?
which numerical scheme are you using now?

Guillaume

**Reply by Raven-pro** [reply]
Hello GiudGiud

Yes, I've worked out the problems using SUPG method. The calculation can get the right solution now.
